### PR TITLE
add_to_nwbfile method added across all data interfaces

### PR DESCRIPTION
**Pull request by bendichter**
The `BaseDataInterface.run_conversion()` method no longer returns an `NWBFile`. Previously, this method returned an `NWBFile` object in many cases, though the behavior was not uniform across DataInterface classes. However, this is not the ideal behavior. When an `nwbfile_path` was passed, the resulting `NWBFile` was bound to a closed HDF5 file, which means none of the Datasets were readable.

Instead, I added a method `BaseDataInterface.create_nwbfile()` which outputs an in-memory `NWBFile` that does not have any bounded datasets. If you want to work with an `NWBFile` in memory, it would be better to use `nwbfile = BaseDataInterface.create_nwbfile()` to create the object and `BaseDataInterface.add_to_nwbfile(nwbfile)` to append to it.

You can still use `BaseDataInterface.run_conversion()`, but this does not return an `NWBFile` object. You can still use `BaseDataInterface.run_conversion(nwbfile=nwbfile)`, in which case the `NWBFile` object is modified in place, however this is deprecated in favor of `BaseDataInterface.add_to_nwbfile(nwbfile=nwbfile)`

### CODE REVIEW

**Comment by bendichter** [comment]
@CodyCBakerPhD this is working for me locally

**Comment by CodyCBakerPhD** [comment]
I'll start working on this and the Zarr backend follow-ups tomorrow

**Comment by CodyCBakerPhD** [comment]
LGTM but it's a big change so extra eyes on this wouldn't hurt

@h-mayorquin Can you take a look as well when you get a chance?

**Review by h-mayorquin — COMMENTED** [review]
This looks good to me. I have some general questions:

> Instead, I added a method BaseDataInterface.create_nwbfile() which outputs an in-memory NWBFile that does not have any bounded datasets. If you want to work with an NWBFile in memory, it would be better to use nwbfile = BaseDataInterface.create_nwbfile() to create the object and BaseDataInterface.add_to_nwbfile(nwbfile) to append to it.

What is the method for writing? Will it remain `run_conversion` or is this something that will be added ina future PR?

I open another issue to discuss something that I thought when I reviewed this PR but that goes beyond the scope here:
https://github.com/catalystneuro/neuroconv/issues/473

File: src/neuroconv/tools/roiextractors/roiextractors.py [hunk 1]
```diff
@@ -970,18 +988,66 @@ def add_summary_images(
...
+    include_roi_centroids: bool = True,
+    include_roi_acceptance: bool = True,
+    mask_type: Optional[str] = "image",  # Literal["image", "pixel"]
+    iterator_options: Optional[dict] = None,
+    compression_options: Optional[dict] = None,
+):
+    # Add device:
+    add_devices(nwbfile=nwbfile, metadata=metadata)
+
+    # ImageSegmentation:
```
[the remark this answers is not in the document]
Comment (h-mayorquin) [reply src/neuroconv/tools/roiextractors/roiextractors.py:1005]: Ah, the old interface has it.

File: src/neuroconv/nwbconverter.py [hunk 2]
```diff
@@ -154,12 +159,7 @@ def run_conversion(
             overwrite=overwrite,
             verbose=self.verbose,
         ) as nwbfile_out:
-            for interface_name, data_interface in self.data_interface_objects.items():
-                data_interface.run_conversion(
-                    nwbfile=nwbfile_out, metadata=metadata, **conversion_options.get(interface_name, dict())
-                )
-
-        return nwbfile_out
+            self.add_to_nwbfile(nwbfile_out, metadata, conversion_options)
```
Comment (h-mayorquin) [inline src/neuroconv/nwbconverter.py:162]: I see, maybe the orchestrators NWBConverter and the Pipeline do have to follow a different logic here.

File: src/neuroconv/tools/neo/neo.py [hunk 3]
```diff
@@ -366,9 +366,9 @@ def add_icephys_recordings(
     # )
 
 
-def add_all_to_nwbfile(
+def add_neo_to_nwb(
```
[the remark this answers is not in the document]
Comment (bendichter) [reply src/neuroconv/tools/neo/neo.py:369]: yes, I feel the naming convention now implies that we are adding these things to the DataInterface object, not an NWBFile.

File: src/neuroconv/basedatainterface.py [hunk 4]
```diff
@@ -39,7 +42,15 @@ def get_metadata(self):
 
         return metadata
 
+    def create_nwbfile(self, metadata=None, **conversion_options):
+        nwbfile = make_nwbfile_from_metadata(metadata)
+        self.add_to_nwbfile(nwbfile, metadata=metadata, **conversion_options)
```
[the remark this answers is not in the document]
Comment (bendichter) [reply src/neuroconv/basedatainterface.py:47]: What would you prefer? Another option could be to rmv `create_nwbfile` and use `make_nwbfile_from_metadata` as a standalone function. That would increase the number of required lines the user needs to write by one line but that might be worth it. So the user would workflow would change from:

```python
some_data_interface.create_nwbfile(nwbfile, metadata=metadata, **conversion_options)
some_data_interface.configure_datasets()
some_data_interface.write(backend=backend, nwbfile=nwbfile, nwbfile_path=nwbfile_path)
```

to

```python
nwbfile = make_nwbfile_from_metadata(metadata)
some_data_interface.add_to_nwbfile(nwbfile, metadata=metadata, **conversion_options)
some_data_interface.configure_datasets()
some_data_interface.write(backend=backend, nwbfile=nwbfile, nwbfile_path=nwbfile_path)
```

I'm OK with that if you think it's clearer.

File: src/neuroconv/tools/neo/neo.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bendichter) [reply src/neuroconv/tools/neo/neo.py:369]: but let's save that for another PR. This one is already pretty big

**Comment by bendichter** [comment]
> What is the method for writing? Will it remain run_conversion or is this something that will be added in a future PR?

Yes, see https://github.com/catalystneuro/neuroconv/issues/441#issuecomment-1544482373 for a rough mock-up

File: src/neuroconv/basedatainterface.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (h-mayorquin) [reply src/neuroconv/basedatainterface.py:47]: I think that the second workflow is better when working with more than one interface.

File: src/neuroconv/tools/neo/neo.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (h-mayorquin) [reply src/neuroconv/tools/neo/neo.py:369]: Makes sense.

**Comment by h-mayorquin** [comment]
> > What is the method for writing? Will it remain run_conversion or is this something that will be added in a future PR?
> 
> Yes, see [#441 (comment)](https://github.com/catalystneuro/neuroconv/issues/441#issuecomment-1544482373) for a rough mock-up

Ah, thanks, great!

File: src/neuroconv/tools/neo/neo.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (CodyCBakerPhD) [reply src/neuroconv/tools/neo/neo.py:369]: Moved to #474

File: src/neuroconv/basedatainterface.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (CodyCBakerPhD) [reply src/neuroconv/basedatainterface.py:47]: If you're working with more than one interface though, it's recommended to just use a `NWBConverter` or `ConverterPipe`, which would have additional capability of metadata aggregation (including combination of schemas)

It seems weird to me to encourage the use of a secondary function in a workflow using interface and converter class instances; even `interface.make_nwbfile_from_metadata()` would be fine to fix my qualm with that (at which point `interface.create_nwbfile(metadata=metadata)` I think communicates the same thing)

The bigger question being asked here, is ' should we support one-liners like `run_conversion` that do TWO operations; (i) making the NWBFile, (ii) populating the NWBFile with all the neurodata types from the streams - or should we break those operations up?

File: src/neuroconv/nwbconverter.py
[the remark this answers is not in the document]
Comment (CodyCBakerPhD) [reply src/neuroconv/nwbconverter.py:1]: Resolved by https://github.com/catalystneuro/neuroconv/pull/455#discussion_r1224296353 I think

**Review by CodyCBakerPhD — APPROVED** [review]

File: src/neuroconv/basedatainterface.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (h-mayorquin) [reply src/neuroconv/basedatainterface.py:47]: Let's accept your framework. From withing it I still think that the create_nwb_method should not also add. The method is doing two things and it is also doing something that it is not very clear from its name. Can we change that and discuss the workflow with multiple interfaces somewhere else?

[the remark this answers is not in the document]
Comment (CodyCBakerPhD) [reply src/neuroconv/basedatainterface.py:47]: Yeah I added a link to it on the main issue https://github.com/catalystneuro/neuroconv/issues/473#issuecomment-1584764451

Feel free to expand/summarize more there

**Pull request merged**